crypto/addresses/src/bech32.rs

```diff
@@ -0,0 +1,195 @@
+use crate::{Address, AddressError, Prefix};
+
+const CHARSET: &[u8] = b"qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+const REV_CHARSET: [u8; 123] = [
+    100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 15, 100, 10, 17, 21, 20, 26, 30, 7, 5, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 100, 29, 100, 24, 13, 25, 9, 8, 23, 100, 18, 22, 31, 27, 19, 100, 1, 0, 3, 16, 11, 28, 12, 14, 6, 4, 2,
+];
+
+// Checksome for bech32
+// https://bch.info/en/specifications
+fn polymod<'data, I>(values: I) -> u64
+where
+    I: Iterator<Item = &'data u8>,
+{
+    let mut c = 1u64;
+    for d in values {
+        let c0 = c >> 35;
+        c = ((c & 0x07ffffffff) << 5) ^ (*d as u64);
+
+        if c0 & 0x01 != 0 {
+            c ^= 0x98f2bc8e61;
+        }
+        if c0 & 0x02 != 0 {
+            c ^= 0x79b76d99e2;
+        }
+        if c0 & 0x04 != 0 {
+            c ^= 0xf33e5fb3c4;
+        }
+        if c0 & 0x08 != 0 {
+            c ^= 0xae2eabe2a8;
+        }
+        if c0 & 0x10 != 0 {
+            c ^= 0x1e4f43e470;
+        }
+    }
+    c ^ 1
+}
+
+fn checksum(payload: &[u8], prefix: &[u8]) -> u64 {
+    polymod(
+        prefix
+            .iter()
+            .chain(&[0u8])
+            .chain(payload)
+            .chain(&vec![0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8]),
+    )
+}
+
+// Convert 8bit array to 5bit array with right padding
+fn conv8to5(payload: &[u8]) -> Vec<u8> {
```

Comment:
Why is this much more complicated than this? https://github.com/kaspanet/kaspad/blob/54fbfadf84ab43cd4eff7df131ebd18b50bdd882/util/bech32/bech32.go#L260
do they do the same thing or something different?

If it's the same it can also return an Iterator instead of collecting to a vec

Reply:
Or is it actually https://github.com/kaspanet/kaspad/blob/54fbfadf84ab43cd4eff7df131ebd18b50bdd882/util/bech32/bech32.go#L166 ?

Reply:
It is the second one. The prefix is done by
```
     let fivebit_prefix: Vec<u8> = self
            .prefix
            .to_string()
            .as_bytes()
            .iter()
            .map(|c| c & 0x1fu8)
            .collect();
```